src/components/CompareResults/beta/DownloadButton.tsx

```diff
@@ -13,10 +16,76 @@ const styles = {
   }),
 };
 
+interface GroupedResults {
+  [key: string]: ResultObject[];
+}
+
+interface ResultObject {
+  header_name: string;
+  new_rev: string;
+}
+
 function DownloadButton() {
+  let fileName = 'perf-compare-all-revisions.json';
+  const results = useAppSelector((state) => {
+    if (
+      state.comparison.activeComparison ===
+      Strings.components.comparisonRevisionDropdown.allRevisions
+    ) {
+      return state.compareResults.data;
+    } else {
+      fileName = `perf-compare-${truncateHash(
+        state.comparison.activeComparison,
+      )}.json`;
+      return {
+        [state.comparison.activeComparison]:
+          state.compareResults.data[state.comparison.activeComparison],
+      };
+    }
+  });
+
+  const formatDownloadData = (
+    data: Record<string, ResultObject[]>,
+  ): object[] => {
+    const groupNames = Object.keys(data);
+    const transformedGroups: object[] = [];
+
+    groupNames.forEach((groupName) => {
+      const groupedResults = data[groupName].reduce(
+        (grouped: GroupedResults, result: ResultObject) => {
+          if (!grouped[result.header_name]) {
+            grouped[result.header_name] = [];
+          }
+          grouped[result.header_name].push(result);
+          return grouped;
+        },
+        {},
+      );
+
+      const transformedGroupEntries = Object.keys(groupedResults).map(
+        (header_name) => ({
+          [`${header_name} ${truncateHash(
+            groupedResults[header_name][0].new_rev,
+          )}`]: groupedResults[header_name],
+        }),
+      );
+
+      transformedGroups.push(...transformedGroupEntries);
+    });
+    return transformedGroups;
+  };
+
   return (
     <div className={styles.downloadButton}>
-      <Button disabled>Download JSON</Button>
+      <Button
+        disabled={!results}
+        href={`data:text/json;charset=utf-8,${encodeURIComponent(
+          JSON.stringify(formatDownloadData(results)),
+        )}`}
```

Comment:
Ideally we should use a Blob and a blob url, that would avoid the need to `encodeURIComponent` and also would avoid putting the full data in the DOM which can have a performance issue if it's big. But it is a bit difficult to get right. Especially it's important to revoke the url and with React it's not super easy, [we have a component just for this in the profiler code](https://github.com/firefox-devtools/profiler/blob/main/src/components/shared/BlobUrlLink.js) that we could reuse... But let's keep it for later.